product-catalog/app/productcatalog/controllers/ProductCatalogController.java

```diff
@@ -1,41 +1,109 @@
-                final ProductOverviewPageContent content = new ProductOverviewPageContent(cms, context(), result, PriceFormatterImpl.of());
-                return render(view -> ok(view.productOverviewPage(content)));
-            })
+                        productService.searchProducts(page, PAGE_SIZE).flatMap(result -> {
+                            final ProductOverviewPageContent content = getPopPageData(cms, result);
+                            return render(view -> ok(view.productOverviewPage(content)));
+                        })
         );
     }
 
-    private SearchDsl<ProductProjection> searchProducts(final int page) {
-        final int offset = (page - 1) * PAGE_SIZE;
-        return ProductProjectionSearch.ofCurrent().withOffset(offset).withLimit(PAGE_SIZE);
+    private ProductOverviewPageContent getPopPageData(final CmsPage cms, final List<ProductProjection> products) {
+        final Translator translator = Translator.of(context().user().language(), context().user().fallbackLanguages(),
+                context().project().languages());
+        final PriceFormatter priceFormatter = PriceFormatter.of(context().user().country().toLocale());
+        final PriceFinder priceFinder = PriceFinder.of(context().user());
+
+        final ProductThumbnailDataFactory thumbnailDataFactory = ProductThumbnailDataFactory.of(translator, priceFinder, priceFormatter);
+
+        final String additionalTitle = "";
+        final List<ProductThumbnailData> productList = products.stream().map(thumbnailDataFactory::create).collect(toList());
+
+        return new ProductOverviewPageContent(additionalTitle, productList);
+    }
+
+    public F.Promise<Result> pdp(final String slug, final String sku) {
+        final F.Promise<Optional<ProductProjection>> productOptPromise = productService.searchProductBySlug(GERMAN, slug);
+
+        return productOptPromise.flatMap(productOptional -> {
+            final Optional<F.Promise<Result>> resultPromise = productOptional.flatMap(product ->
+                    productService.findVariantBySku(product, sku).map(variant -> pdpx(product, variant)));
+            return resultPromise.orElse(F.Promise.pure(notFound()));
+        });
+    }
+
+    private F.Promise<Result> pdpx(final ProductProjection product, final ProductVariant variant) {
+        final F.Promise<List<ProductProjection>> suggestionPromise = productService.getSuggestions(categoryService.getSiblingCategories(product), NUM_SUGGESTIONS);
+        final List<RichShippingRate> shippingRates = shippingMethodService.getShippingRates(context().user().zone());
+        final List<Category> breadcrumbs = categoryService.getBreadCrumbCategories(product);
+
+        return suggestionPromise.flatMap(suggestions -> withCms("pdp", cms -> {
+            final Translator translator = Translator.of(context().user().language(), context().user().fallbackLanguages(), context().project().languages());
```

Comment:
I would extract a method `getTranslator()` since you only use context to initialize it.